source/extensions/stat_sinks/hystrix/hystrix.cc

```diff
@@ -0,0 +1,372 @@
+#include "extensions/stat_sinks/hystrix/hystrix.h"
+
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <sstream>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+#include "common/http/headers.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace StatSinks {
+namespace Hystrix {
+
+const uint64_t HystrixSink::DEFAULT_NUM_OF_BUCKETS;
+
+ClusterStatsCache::ClusterStatsCache(const std::string& cluster_name)
+    : cluster_name_(cluster_name) {}
+
+void ClusterStatsCache::printToStream(std::stringstream& out_str) {
+  const std::string cluster_name_with_prefix = absl::StrCat(cluster_name_, ".");
+
+  printRollingWindow(absl::StrCat(cluster_name_with_prefix, "errors"), success_, out_str);
+  printRollingWindow(absl::StrCat(cluster_name_with_prefix, "success"), errors_, out_str);
+  printRollingWindow(absl::StrCat(cluster_name_with_prefix, "total"), timeouts_, out_str);
+  printRollingWindow(absl::StrCat(cluster_name_with_prefix, "timeouts"), rejected_, out_str);
+  printRollingWindow(absl::StrCat(cluster_name_with_prefix, "rejected"), total_, out_str);
+}
+
+void ClusterStatsCache::printRollingWindow(absl::string_view name, RollingWindow rolling_window,
+                                           std::stringstream& out_str) {
+  out_str << name << " | ";
+  for (auto specific_stat_vec_itr = rolling_window.begin();
+       specific_stat_vec_itr != rolling_window.end(); ++specific_stat_vec_itr) {
+    out_str << *specific_stat_vec_itr << " | ";
+  }
+  out_str << std::endl;
+}
+
+// Add new value to rolling window, in place of oldest one.
+void HystrixSink::pushNewValue(RollingWindow& rolling_window, uint64_t value) {
+  if (rolling_window.empty()) {
+    rolling_window.resize(window_size_, value);
+  } else {
+    rolling_window[current_index_] = value;
+  }
+}
+
+uint64_t HystrixSink::getRollingValue(RollingWindow rolling_window) {
+
+  if (rolling_window.empty()) {
+    return 0;
+  }
+  // If the counter was reset, the result is negative
+  // better return 0, will be back to normal once one rolling window passes.
+  if (rolling_window[current_index_] < rolling_window[(current_index_ + 1) % window_size_]) {
+    return 0;
+  } else {
+    return rolling_window[current_index_] - rolling_window[(current_index_ + 1) % window_size_];
+  }
+}
+
+void HystrixSink::updateRollingWindowMap(Upstream::ClusterInfoConstSharedPtr cluster_info,
```

Comment:
nit: can we just pass this as a `const ClusterInfo&` (the result of `*cluster_info`) to save a shared_ptr copy?